idaes/apps/uncertainty_propagation/sens.py

```diff
@@ -0,0 +1,731 @@
+        If no constraint exists, return []
+    col: list
+        list of variable names
+    line_dic: dict
+        column numbers of the theta_names in the model. Index starts from 1
+
+    Raises
+    ------
+    RuntimeError
+        When ipopt or kaug or dotsens is not available
+    Exception
+        When ipopt fails 
+    """
+    #Create the solver plugin using the ASL interface
+    ipopt = SolverFactory('ipopt',solver_io='nl')
+    if solver_options is not None:
+        ipopt.options = solver_options
+    kaug = SolverFactory('k_aug',solver_io='nl')
+    dotsens = SolverFactory('dot_sens',solver_io='nl')
+    if not ipopt.available(False):
+        raise RuntimeError('ipopt is not available')
+    if not kaug.available(False):
+        raise RuntimeError('k_aug is not available')
+    if not dotsens.available(False):
+        raise RuntimeError('dotsens is not available')
+
+    # Declare Suffixes
+    _add_sensitivity_suffixes(model)
+
+    # K_AUG SUFFIXES
+    model.dof_v = Suffix(direction=Suffix.EXPORT)  #: SUFFIX FOR K_AUG
+    model.rh_name = Suffix(direction=Suffix.IMPORT)  #: SUFFIX FOR K_AUG AS WELL
+    kaug.options["print_kkt"] = ""
+    results = ipopt.solve(model,tee=tee)
+
+    # Rasie Exception if ipopt fails 
+    if (results.solver.status == SolverStatus.warning):
+        raise Exception(results.solver.Message)
+
+    for o in model.component_objects(Objective, active=True):
+        f_mean = value(o)
+    model.ipopt_zL_in.update(model.ipopt_zL_out)
+    model.ipopt_zU_in.update(model.ipopt_zU_out)
+    #: run k_aug
+    kaug.solve(model, tee=tee)  #: always call k_aug AFTER ipopt.
+    model.write('col_row.nl', format='nl', io_options={'symbolic_solver_labels':True})
+    # get the column numbers of theta
+    line_dic = {}
+    try:
+        for v in theta_names:
+            line_dic[v] = line_num('col_row.col', v)
+        # load gradient of the objective function
+        gradient_f = np.loadtxt("./GJH/gradient_f_print.txt")
+        with open ("col_row.col", "r") as myfile:
+            col = myfile.read().splitlines()
+    except Exception as e:
+        print('File not found.')
+        raise e
+    '''
+    gradient_f_dic = {}
+    for i in range(len(col)):
+        gradient_f_dic["d(f)/d("+col[i]+")"] = gradient_f[i]
+    '''
+    # load gradient of all constraints (sparse)
+    # If no constraint exists, return []
+    num_constraints = len(list(model.component_data_objects(Constraint,
+                                                            active=True,
+                                                            descend_into=True)))
+    if num_constraints > 0 :
+        try:
+            gradient_c = np.loadtxt("./GJH/A_print.txt")
+        except Exception as e:
+            print('./GJH/A_print.txt not found.')
+        gradient_c = np.array([i for i in gradient_c if not np.isclose(i[2],0)])
```

Comment:
Make the tolerance for np.isclose an optional input for this function. This will alter the user that we are approximately what is a zero.

Reply:
What is the returned datatype of `gradient_c`? Is it a sparse numpy array? What are the dimensions? I think the dimensions should be Ncon by (Nx + Ntheta) or Ncon by Nx (depending on if x contains theta or not). Please add the dimensions to the doc string and then check they are correct. When in doubt, check that the dimensions work with the main formulas given in the documentation.

Reply:
- added `atol=0.0 into np.isclose.
- I added dimension. Here, I wanted to provide the same format in `k_aug`. In other words, [ [col, row, val], [col, row, val],..,[col, row, val]] index starts from1. So, I use the numpy array. I converted it into sparse in idase toolbox.